notmuch.1
.\" notmuch - Not much of an email program, (just index, search and tagging)
.\"
.\" Copyright © 2009 Carl Worth
.\"
.\" Notmuch is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" Notmuch is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see http://www.gnu.org/licenses/ .
.\"
.\" Author: Carl Worth <cworth@cworth.org>
.TH NOTMUCH 1 2009-10-31 "Notmuch 0.1"
.SH NAME
notmuch \- thread-based email index, search, and tagging
.SH SYNOPSIS
.B notmuch
.IR command " [" args " ...]"
.SH DESCRIPTION
Notmuch is a command-line based program for indexing, searching,
reading, and tagging large collections of email messages.

The quickest way to get started with Notmuch is to simply invoke the
.B notmuch
command with no arguments, which will interactively guide you through
the process of indexing your mail.
.SH NOTE
While the command-line program
.B notmuch
provides powerful functionality, it does not provide the most
convenient interface for that functionality. More sophisticated
interfaces are expected to be built on top of either the command-line
interface, or more likely, on top of the notmuch library
interface. See http://notmuchmail.org for more about alternate
interfaces to notmuch.
.SH COMMANDS
The
.BR setup
command is used to configure Notmuch for first use, (or to reconfigure
it later).
.RS 4
.TP 4
.B setup

Interactively sets up notmuch for first use.

The setup command will prompt for your full name, your primary email
address, any alternate email addresses you use, and the directory
containing your email archives. Your answers will be written to a
configuration file in ${NOTMUCH_CONFIG} (if set) or
${HOME}/.notmuch-config . This configuration file will be created with
descriptive comments, making it easy to edit by hand later to change the
configuration. Or you can run
.B "notmuch setup"
again to change the configuration.

The mail directory you specify can contain any number of
sub-directories and should primarily contain only files with individual
email messages (eg. maildir or mh archives are perfect). If there are
other, non-email files (such as indexes maintained by other email
programs) then notmuch will do its best to detect those and ignore
them.

Mail storage that uses mbox format, (where one mbox file contains many
messages), will not work with notmuch. If that's how your mail is
currently stored, it is recommended you first convert it to maildir
format with a utility such as mb2md before running
.B "notmuch setup" .

Invoking
.B notmuch
with no command argument will run
.B setup
if the setup command has not previously been completed.
.RE

The
.B new
command is used to incorporate new mail into the notmuch database.
.RS 4
.TP 4
.B new

Find and import any new messages to the database.

The
.B new
command scans all sub-directories of the database, performing
full-text indexing on new messages that are found. Each new message
will automatically be tagged with the
.BR inbox
tag, and, unless it was already "seen" by another client, the
.BR unread
tag.

You should run
.B "notmuch new"
once after first running
.B "notmuch setup"
to create the initial database. The first run may take a long time if
you have a significant amount of mail (several hundred thousand
messages or more). Subsequently, you should run
.B "notmuch new"
whenever new mail is delivered and you wish to incorporate it into the
database. These subsequent runs will be much quicker than the initial
run.

Invoking
.B notmuch
with no command argument will run
.B new
if
.B "notmuch setup"
has previously been completed, but
.B "notmuch new"
has not previously been run.
.RE

Several of the notmuch commands accept search terms with a common
syntax. See the
.B "SEARCH SYNTAX"
section below for more details on the supported syntax.

The
.BR search ", " show " and " count
commands are used to query the email database.
.RS 4
.TP 4
.BR search " [options...] <search-term>..."

Search for messages matching the given search terms, and display as
results the threads containing the matched messages.

The output consists of one line per thread, giving a thread ID, the
date of the newest (or oldest, depending on the sort option) matched
message in the thread, the number of matched messages and total
messages in the thread, the names of all participants in the thread,
and the subject of the newest (or oldest) message.

Supported options for
.B search
include
.RS 4
.TP 4
.BR \-\-format= ( json | text )

Presents the results in either JSON or plain-text (default).
.RE
.RS 4
.TP 4
.BR \-\-sort= ( newest\-first | oldest\-first )

This option can be used to present results in either chronological order
.RB ( oldest\-first )
or reverse chronological order
.RB ( newest\-first ).

Note: The thread order will be distinct between these two options
(beyond being simply reversed). When sorting by
.B oldest\-first
the threads will be sorted by the oldest message in each thread, but
when sorting by
.B newest\-first
the threads will be sorted by the newest message in each thread.

.RE
.RS 4
By default, results will be displayed in reverse chronological order,
(that is, the newest results will be displayed first).

See the
.B "SEARCH SYNTAX"
section below for details of the supported syntax for <search-terms>.
.RE
.TP
.BR show " [options...] <search-term>..."

Shows all messages matching the search terms.

The messages will be grouped and sorted based on the threading (all
replies to a particular message will appear immediately after that
message in date order). The output is not indented by default, but
depth tags are printed so that proper indentation can be performed by
a post-processor (such as the emacs interface to notmuch).

Supported options for
.B show
include
.RS 4
.TP 4
.B \-\-entire\-thread

By default only those messages that match the search terms will be
displayed. With this option, all messages in the same thread as any
matched message will be displayed.
.RE

.RS 4
.TP 4
.B \-\-format=(text|json|mbox)

.RS 4
.TP 4
.B text

The default plain-text format has all text-content MIME parts
decoded. Various components in the output,
.RB ( message ", " header ", " body ", " attachment ", and MIME " part ),
will be delimited by easily-parsed markers. Each marker consists of a
Control-L character (ASCII decimal 12), the name of the marker, and
then either an opening or closing brace, ('{' or '}'), to either open
or close the component.

.RE
.RS 4
.TP 4
.B json

The output is formatted with Javascript Object Notation (JSON). This
format is more robust than the text format for automated
processing. JSON output always includes all messages in a matching
thread; in effect
.B \-\-format=json
implies
.B \-\-entire\-thread

.RE
.RS 4
.TP 4
.B mbox

All matching messages are output in the traditional, Unix mbox format
with each message being prefixed by a line beginning with "From " and
a blank line separating each message. Lines in the message content
beginning with "From " (preceded by zero or more '>' characters) have
an additional '>' character added. This reversible escaping
is termed "mboxrd" format and described in detail here:
http://homepage.ntlworld.com/jonathan.deboynepollard/FGA/mail-mbox-formats.html

.RE
A common use of
.B notmuch show
is to display a single thread of email messages. For this, use a
search term of "thread:<thread-id>" as can be seen in the first
column of output from the
.B notmuch search
command.

See the
.B "SEARCH SYNTAX"
section below for details of the supported syntax for <search-terms>.
.RE
.TP
.BR count " <search-term>..."

Count messages matching the search terms.

The number of matching messages is output to stdout.

With no search terms, a count of all messages in the database will be
displayed.
.RE
.RE

The
.B reply
command is useful for preparing a template for an email reply.
.RS 4
.TP 4
.BR reply " [options...] <search-term>..."

Constructs a reply template for a set of messages.

To make replying to email easier,
.B notmuch reply
takes an existing set of messages and constructs a suitable mail
template. The Reply-to header (if any, otherwise From:) is used for
the To: address. Vales from the To: and Cc: headers are copied, but
not including any of the current user's email addresses (as configured
in primary_mail or other_email in the .notmuch\-config file) in the
recipient list

It also builds a suitable new subject, including Re: at the front (if
not already present), and adding the message IDs of the messages being
replied to to the References list and setting the In\-Reply\-To: field
correctly.

Finally, the original contents of the emails are quoted by prefixing
each line with '> ' and included in the body.

The resulting message template is output to stdout.

Supported options for
.B reply
include
.RS
.TP 4
.BR \-\-format= ( default | headers\-only )
.RS
.TP 4
.BR default
Includes subject and quoted message body.
.TP
.BR headers\-only
Only produces In\-Reply\-To, References, To, Cc, and Bcc headers.
.RE

See the
.B "SEARCH SYNTAX"
section below for details of the supported syntax for <search-terms>.

Note: It is most common to use
.B "notmuch reply"
with a search string matching a single message, (such as
id:<message-id>), but it can be useful to reply to several messages at
once. For example, when a series of patches are sent in a single
thread, replying to the entire thread allows for the reply to comment
on issue found in multiple patches.
.RE
.RE

The
.B tag
command is the only command available for manipulating database
contents.

.RS 4
.TP 4
.BR tag " +<tag>|\-<tag> [...] [\-\-] <search-term>..."

Add/remove tags for all messages matching the search terms.

Tags prefixed by '+' are added while those prefixed by '\-' are
removed. For each message, tag removal is performed before tag
addition.

The beginning of <search-terms> is recognized by the first
argument that begins with neither '+' nor '\-'. Support for
an initial search term beginning with '+' or '\-' is provided
by allowing the user to specify a "\-\-" argument to separate
the tags from the search terms.

See the
.B "SEARCH SYNTAX"
section below for details of the supported syntax for <search-terms>.
.RE

The
.BR dump " and " restore
commands can be used to create a textual dump of email tags for backup
purposes, and to restore from that dump

.RS 4
.TP 4
.BR dump " [<filename>]"

Creates a plain-text dump of the tags of each message.

The output is to the given filename, if any, or to stdout.

These tags are the only data in the notmuch database that can't be
recreated from the messages themselves.  The output of notmuch dump is
therefore the only critical thing to backup (and much more friendly to
incremental backup than the native database files.)
.TP
.BR restore " <filename>"

Restores the tags from the given file (see
.BR "notmuch dump" "."

Note: The dump file format is specifically chosen to be
compatible with the format of files produced by sup-dump.
So if you've previously been using sup for mail, then the
.B "notmuch restore"
command provides you a way to import all of your tags (or labels as
sup calls them).
.RE

The
.B part
command can used to output a single part of a multi-part MIME message.

.RS 4
.TP 4
.BR part " \-\-part=<part-number> <search-term>..."

Output a single MIME part of a message.

A single decoded MIME part, with no encoding or framing, is output to
stdout. The search terms must match only a single message, otherwise
this command will fail.

The part number should match the part "id" field output by the
"\-\-format=json" option of "notmuch show". If the message specified by
the search terms does not include a part with the specified "id" there
will be no output.

See the
.B "SEARCH SYNTAX"
section below for details of the supported syntax for <search-terms>.
.RE

.SH SEARCH SYNTAX
Several notmuch commands accept a common syntax for search terms.

The search terms can consist of free-form text (and quoted phrases)
which will match all messages that contain all of the given
terms/phrases in the body, the subject, or any of the sender or
recipient headers.

As a special case, a search string consisting of exactly a single
asterisk ("*") will match all messages.

In addition to free text, the following prefixes can be used to force
terms to match against specific portions of an email, (where
<brackets> indicate user-supplied values):

	from:<name-or-address>

	to:<name-or-address>

	subject:<word-or-quoted-phrase>

	attachment:<word>

	tag:<tag> (or is:<tag>)

	id:<message-id>

	thread:<thread-id>

The
.B from:
prefix is used to match the name or address of the sender of an email
message.

The
.B to:
prefix is used to match the names or addresses of any recipient of an
email message, (whether To, Cc, or Bcc).

Any term prefixed with
.B subject:
will match only text from the subject of an email. Searching for a
phrase in the subject is supported by including quotation marks around
the phrase, immediately following
.BR subject: .

The
.B attachment:
prefix can be used to search for specific filenames (or extensions) of
attachments to email messages.

For
.BR tag: " and " is:
valid tag values include
.BR inbox " and " unread
by default for new messages added by
.B notmuch new
as well as any other tag values added manually with
.BR "notmuch tag" .

For
.BR id: ,
message ID values are the literal contents of the Message\-ID: header
of email messages, but without the '<', '>' delimiters.

The
.B thread:
prefix can be used with the thread ID values that are generated
internally by notmuch (and do not appear in email messages). These
thread ID values can be seen in the first column of output from
.B "notmuch search"

In addition to individual terms, multiple terms can be
combined with Boolean operators (
.BR and ", " or ", " not
, etc.). Each term in the query will be implicitly connected by a
logical AND if no explicit operator is provided, (except that terms
with a common prefix will be implicitly combined with OR until we get
Xapian defect #402 fixed).

Parentheses can also be used to control the combination of the Boolean
operators, but will have to be protected from interpretation by the
shell, (such as by putting quotation marks around any parenthesized
expression).

Finally, results can be restricted to only messages within a
particular time range, (based on the Date: header) with a syntax of:

	<intial-timestamp>..<final-timestamp>

Each timestamp is a number representing the number of seconds since
1970\-01\-01 00:00:00 UTC. This is not the most convenient means of
expressing date ranges, but until notmuch is fixed to accept a more
convenient form, one can use the date program to construct
timestamps. For example, with the bash shell the folowing syntax would
specify a date range to return messages from 2009\-10\-01 until the
current time:

	$(date +%s \-d 2009\-10\-01)..$(date +%s)
.SH ENVIRONMENT
The following environment variables can be used to control the
behavior of notmuch.
.TP
.B NOTMUCH_CONFIG

notmuch.c
/* notmuch - Not much of an email program, (just index and search)
 *
 * Copyright © 2009 Carl Worth
 * Copyright © 2009 Keith Packard
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/ .
 *
 * Authors: Carl Worth <cworth@cworth.org>
 *	    Keith Packard <keithp@keithp.com>
 */

#include "notmuch-client.h"

typedef int (*command_function_t) (void *ctx, int argc, char *argv[]);

typedef struct command {
    const char *name;
    command_function_t function;
    const char *arguments;
    const char *summary;
    const char *documentation;
} command_t;

static int
notmuch_help_command (void *ctx, int argc, char *argv[]);

static const char search_terms_help[] =
    "\tSeveral notmuch commands accept a comman syntax for search\n"
    "\tterms.\n"
    "\n"
    "\tThe search terms can consist of free-form text (and quoted\n"
    "\tphrases) which will match all messages that contain all of\n"
    "\tthe given terms/phrases in the body, the subject, or any of\n"
    "\tthe sender or recipient headers.\n"
    "\n"
    "\tAs a special case, a search string consisting of exactly a\n"
    "\tsingle asterisk (\"*\") will match all messages.\n"
    "\n"
    "\tIn addition to free text, the following prefixes can be used\n"
    "\tto force terms to match against specific portions of an email,\n"
    "\t(where <brackets> indicate user-supplied values):\n"
    "\n"
    "\t\tfrom:<name-or-address>\n"
    "\t\tto:<name-or-address>\n"
    "\t\tsubject:<word-or-quoted-phrase>\n"
    "\t\tattachment:<word>\n"
    "\t\ttag:<tag> (or is:<tag>)\n"
    "\t\tid:<message-id>\n"
    "\t\tthread:<thread-id>\n"
    "\n"
    "\tThe from: prefix is used to match the name or address of\n"
    "\tthe sender of an email message.\n"
    "\n"
    "\tThe to: prefix is used to match the names or addresses of\n"
    "\tany recipient of an email message, (whether To, Cc, or Bcc).\n"
    "\n"
    "\tAny term prefixed with subject: will match only text from\n"
    "\tthe subject of an email. Quoted phrases are supported when\n"
    "\tsearching with: subject:\"this is a phrase\".\n"
    "\n"
    "\tFor tag: and is:, valid tag values include \"inbox\" and \"unread\"\n"
    "\tby default for new messages added by \"notmuch new\" as well\n"
    "\tas any other tag values added manually with \"notmuch tag\".\n"
    "\n"
    "\tFor id:, message ID values are the literal contents of the\n"
    "\tMessage-ID: header of email messages, but without the '<','>'\n"
    "\tdelimiters.\n"
    "\n"
    "\tThe thread: prefix can be used with the thread ID values that\n"
    "\tare generated internally by notmuch (and do not appear in email\n"
    "\tmessages). These thread ID values can be seen in the first\n"
    "\tcolumn of output from \"notmuch search\".\n"
    "\n"
    "\tIn addition to individual terms, multiple terms can be\n"
    "\tcombined with Boolean operators (\"and\", \"or\", \"not\", etc.).\n"
    "\tEach term in the query will be implicitly connected by a\n"
    "\tlogical AND if no explicit operator is provided, (except\n"
    "\tthat terms with a common prefix will be implicitly combined\n"
    "\twith OR until we get Xapian defect #402 fixed).\n"
    "\n"
    "\tParentheses can also be used to control the combination of\n"
    "\tthe Boolean operators, but will have to be protected from\n"
    "\tinterpretation by the shell, (such as by putting quotation\n"
    "\tmarks around any parenthesized expression).\n"
    "\n"
    "\tFinally, results can be restricted to only messages within a\n"
    "\tparticular time range, (based on the Date: header) with:\n"
    "\n"
    "\t\t<intial-timestamp>..<final-timestamp>\n"
    "\n"
    "\tEach timestamp is a number representing the number of seconds\n"
    "\tsince 1970-01-01 00:00:00 UTC. This is not the most convenient\n"
    "\tmeans of expressing date ranges, but until notmuch is fixed to\n"
    "\taccept a more convenient form, one can use the date program to\n"
    "\tconstruct timestamps. For example, with the bash shell the\n"
    "\tfollowing syntax would specify a date range to return messages\n"
    "\tfrom 2009-10-01 until the current time:\n"
    "\n"
    "\t\t$(date +%%s -d 2009-10-01)..$(date +%%s)\n\n";

command_t commands[] = {
    { "setup", notmuch_setup_command,
      NULL,
      "Interactively setup notmuch for first use.",
      "\tThe setup command will prompt for your full name, your primary\n"
      "\temail address, any alternate email addresses you use, and the\n"
      "\tdirectory containing your email archives. Your answers will be\n"
      "\twritten to a configuration file in ${NOTMUCH_CONFIG} (if set)\n"
      "\tor ${HOME}/.notmuch-config.\n"
      "\n"
      "\tThis configuration file will be created with descriptive\n"
      "\tcomments, making it easy to edit by hand later to change the\n"
      "\tconfiguration. Or you can run \"notmuch setup\" again.\n"
      "\n"
      "\tInvoking notmuch with no command argument will run setup if\n"
      "\tthe setup command has not previously been completed." },
    { "new", notmuch_new_command,
      "[--verbose]",
      "Find and import new messages to the notmuch database.",
      "\t\tScans all sub-directories of the mail directory, performing\n"
      "\t\tfull-text indexing on new messages that are found. Each new\n"
      "\t\tmessage will be tagged as \"inbox\" and, unless it is\n"
      "\t\tmarked as \"seen\", \"unread\".\n"
      "\n"
      "\tYou should run \"notmuch new\" once after first running\n"
      "\t\"notmuch setup\" to create the initial database. The first\n"
      "\trun may take a long time if you have a significant amount of\n"
      "\tmail (several hundred thousand messages or more).\n"
      "\n"
      "\tSubsequently, you should run \"notmuch new\" whenever new mail\n"
      "\tis delivered and you wish to incorporate it into the database.\n"
      "\tThese subsequent runs will be much quicker than the initial run.\n"
      "\n"
      "\tSupported options for new include:\n"
      "\n"
      "\t--verbose\n"
      "\n"
      "\t\tVerbose operation. Shows paths of message files as\n"
      "\t\tthey are being indexed.\n"
      "\n"
      "\tInvoking notmuch with no command argument will run new if\n"
      "\tthe setup command has previously been completed, but new has\n"
      "\tnot previously been run." },
    { "search", notmuch_search_command,
      "[options...] <search-terms> [...]",
      "Search for messages matching the given search terms.",
      "\tNote that the individual mail messages will be matched\n"
      "\tagainst the search terms, but the results will be the\n"
      "\tthreads (one per line) containing the matched messages.\n"
      "\n"
      "\tSupported options for search include:\n"
      "\n"
      "\t--format=(json|text)\n"
      "\n"
      "\t\tPresents the results in either JSON or\n"
      "\t\tplain-text (default)\n"
      "\n"
      "\t--sort=(newest-first|oldest-first)\n"
      "\n"
      "\t\tPresent results in either chronological order\n"
      "\t\t(oldest-first) or reverse chronological order\n"
      "\t\t(newest-first), which is the default.\n"
      "\n"
      "\tSee \"notmuch help search-terms\" for details of the search\n"
      "\tterms syntax." },
    { "show", notmuch_show_command,
      "<search-terms> [...]",
      "Show all messages matching the search terms.",
      "\tThe messages are grouped and sorted based on the threading\n"
      "\t(all replies to a particular message appear immediately\n"
      "\tafter that message in date order).\n"
      "\n"
      "\tSupported options for show include:\n"
      "\n"
      "\t--entire-thread\n"
      "\n"
      "\t\tBy default only those messages that match the\n"
      "\t\tsearch terms will be displayed. With this option,\n"
      "\t\tall messages in the same thread as any matched\n"
      "\t\tmessage will be displayed.\n"
      "\n"
      "\t--format=(text|json|mbox)\n"
      "\n"
      "\t\ttext (default)\n"
      "\n"
      "\t\tThe default plain-text format has all text-content MIME parts\n"
      "\t\tdecoded. Various components in the output, ('message', 'header',\n"
      "\t\t'body', 'attachment', and MIME 'part') are delimited by\n"
      "\t\teasily-parsed markers. Each marker consists of a Control-L\n"
      "\t\tcharacter (ASCII decimal 12), the name of the marker, and\n"
      "\t\tthen either an opening or closing brace, '{' or '}' to\n"
      "\t\teither open or close the component.\n"
      "\n"
      "\t\tjson\n"
      "\n"
      "\t\tThe output is formatted with Javascript Object Notation\n"
      "\t\t(JSON). This format is more robust than the text format\n"
      "\t\tfor automated processing. JSON output always includes all\n"
      "\t\tmessages in a matching thread; in effect '--format=json'\n"
      "\t\timplies '--entire-thread'\n"
      "\n"
      "\t\tmbox\n"
      "\n"
      "\t\tAll matching messages are output in the traditional, Unix\n"
      "\t\tmbox format with each message being prefixed by a line\n"
      "\t\tbeginning with 'From ' and a blank line separating each\n"
      "\t\tmessage. Lines in the message content beginning with 'From '\n"
      "\t\t(preceded by zero or more '>' characters) have an additional\n"
      "\t\t'>' character added. This reversible escaping is termed\n"
      "\t\t\"mboxrd\" format and described in detail here:\n"
      "\n"
      "\t\thttp://homepage.ntlworld.com/jonathan.deboynepollard/FGA/mail-mbox-formats.html\n"
      "\n"
      "\tA common use of \"notmuch show\" is to display a single\n"
      "\tthread of email messages. For this, use a search term of\n"
      "\t\"thread:<thread-id>\" as can be seen in the first column\n"
      "\tof output from the \"notmuch search\" command.\n"
      "\n"
      "\tSee \"notmuch help search-terms\" for details of the search\n"
      "\tterms syntax." },
    { "count", notmuch_count_command,
      "<search-terms> [...]",
      "Count messages matching the search terms.",
      "\tThe number of matching messages is output to stdout.\n"
      "\n"
      "\tWith no search terms, a count of all messages in the database\n"
      "\twill be displayed.\n"
      "\n"
      "\tSee \"notmuch help search-terms\" for details of the search\n"
      "\tterms syntax." },
    { "reply", notmuch_reply_command,
      "[options...] <search-terms> [...]",
      "Construct a reply template for a set of messages.",
      "\tConstructs a new message as a reply to a set of existing\n"
      "\tmessages. The Reply-To: header (if any, otherwise From:) is\n"
      "\tused for the To: address. The To: and Cc: headers are copied,\n"
      "\tbut not including any of the user's configured addresses.\n"
      "\n"
      "\tA suitable subject is constructed. The In-Reply-to: and\n"
      "\tReferences: headers are set appropriately, and the content\n"
      "\tof the original messages is quoted and included in the body\n"
      "\t(unless --format=headers-only is given).\n"
      "\n"
      "\tThe resulting message template is output to stdout.\n"
      "\n"
      "\tSupported options for reply include:\n"
      "\n"
      "\t--format=(default|headers-only)\n"
      "\n"
      "\t\tdefault:\n"
      "\t\t\tIncludes subject and quoted message body.\n"
      "\n"
      "\t\theaders-only:\n"
      "\t\t\tOnly produces In-Reply-To, References, To\n"
      "\t\t\tCc, and Bcc headers.\n"
      "\n"
      "\tSee \"notmuch help search-terms\" for details of the search\n"
      "\tterms syntax." },
    { "tag", notmuch_tag_command,
      "+<tag>|-<tag> [...] [--] <search-terms> [...]",
      "Add/remove tags for all messages matching the search terms.",
      "\tThe search terms are handled exactly as in 'search' so one\n"
      "\tcan use that command first to see what will be modified.\n"
      "\n"
      "\tTags prefixed by '+' are added while those prefixed by\n"
      "\t'-' are removed. For each message, tag removal is performed\n"
      "\tbefore tag addition.\n"
      "\n"
      "\tThe beginning of <search-terms> is recognized by the first\n"
      "\targument that begins with neither '+' nor '-'. Support for\n"
      "\tan initial search term beginning with '+' or '-' is provided\n"
      "\tby allowing the user to specify a \"--\" argument to separate\n"
      "\tthe tags from the search terms.\n"
      "\n"
      "\tSee \"notmuch help search-terms\" for details of the search\n"
      "\tterms syntax." },
    { "dump", notmuch_dump_command,
      "[<filename>]",
      "Create a plain-text dump of the tags for each message.",
      "\tOutput is to the given filename, if any, or to stdout.\n"
      "\tThese tags are the only data in the notmuch database\n"
      "\tthat can't be recreated from the messages themselves.\n"
      "\tThe output of notmuch dump is therefore the only\n"
      "\tcritical thing to backup (and much more friendly to\n"
      "\tincremental backup than the native database files.)" },
    { "restore", notmuch_restore_command,
      "<filename>",
      "Restore the tags from the given dump file (see 'dump').",
      "\tNote: The dump file format is specifically chosen to be\n"
      "\tcompatible with the format of files produced by sup-dump.\n"
      "\tSo if you've previously been using sup for mail, then the\n"
      "\t\"notmuch restore\" command provides you a way to import\n"
      "\tall of your tags (or labels as sup calls them)." },
    { "search-tags", notmuch_search_tags_command,
      "[<search-terms> [...] ]",
      "List all tags found in the database or matching messages.",
      "\tRun this command without any search-term(s) to obtain a list\n"
      "\tof all tags found in the database. If you provide one or more\n"
      "\tsearch-terms as argument(s) then the resulting list will\n"
      "\tcontain tags only from messages that match the search-term(s).\n"
      "\n"
      "\tIn both cases the list will be alphabetically sorted." },
    { "part", notmuch_part_command,
      "--part=<num> <search-terms>",
      "Output a single MIME part of a message.",
      "\tA single decoded MIME part, with no encoding or framing,\n"
      "\tis output to stdout. The search terms must match only a single\n"
      "\tmessage, otherwise this command will fail.\n"
      "\n"
      "\tThe part number should match the part \"id\" field output\n"
      "\tby the \"--format=json\" option of \"notmuch show\". If the\n"
      "\tmessage specified by the search terms does not include a\n"
      "\tpart with the specified \"id\" there will be no output." },
    { "help", notmuch_help_command,
      "[<command>]",
      "This message, or more detailed help for the named command.",
      "\tExcept in this case, where there's not much more detailed\n"
      "\thelp available." }
};

static void
usage (FILE *out)
{
    command_t *command;
    unsigned int i;

    fprintf (out,
	     "Usage: notmuch --help\n"
	     "       notmuch --version\n"
	     "       notmuch <command> [args...]\n");
    fprintf (out, "\n");
    fprintf (out, "The available commands are as follows:\n");
    fprintf (out, "\n");

    for (i = 0; i < ARRAY_SIZE (commands); i++) {
	command = &commands[i];

	fprintf (out, "  %-11s  %s\n",
		 command->name, command->summary);
    }

    fprintf (out, "\n");
    fprintf (out,
    "Use \"notmuch help <command>\" for more details on each command\n"
    "and \"notmuch help search-terms\" for the common search-terms syntax.\n\n");
}

static int
notmuch_help_command (unused (void *ctx), int argc, char *argv[])
{
    command_t *command;
    unsigned int i;

    if (argc == 0) {
	printf ("The notmuch mail system.\n\n");
	usage (stdout);
	return 0;
    }

    for (i = 0; i < ARRAY_SIZE (commands); i++) {
	command = &commands[i];

	if (strcmp (argv[0], command->name) == 0) {
	    printf ("Help for \"notmuch %s\":\n\n", argv[0]);
	    if (command->arguments)
		printf ("%s %s\n\n\t%s\n\n%s\n\n",
			command->name, command->arguments,
			command->summary, command->documentation);
	    else
		printf ("%s\t%s\n\n%s\n\n", command->name,
			command->summary, command->documentation);
	    return 0;
	}
    }

    if (strcmp (argv[0], "search-terms") == 0) {
	printf ("Help for <%s>\n\n", argv[0]);
	for (i = 0; i < ARRAY_SIZE (commands); i++) {
	    command = &commands[i];

	    if (command->arguments &&
		strstr (command->arguments, "search-terms"))
	    {
		printf ("\t%s\t%s\n",
			command->name, command->arguments);
	    }
	}
	printf ("\n");
	printf (search_terms_help);
	return 0;
    }

    fprintf (stderr,
	     "\nSorry, %s is not a known command. There's not much I can do to help.\n\n",
	     argv[0]);
    return 1;
}

/* Handle the case of "notmuch" being invoked with no command
 * argument. For now we just call notmuch_setup_command, but we plan
 * to be more clever about this in the future.
 */
static int
notmuch (void *ctx)
{
    notmuch_config_t *config;
    notmuch_bool_t is_new;
    char *db_path;
    struct stat st;

    config = notmuch_config_open (ctx, NULL, &is_new);

    /* If the user has never configured notmuch, then run
     * notmuch_setup_command which will give a nice welcome message,
     * and interactively guide the user through the configuration. */
    if (is_new) {
	notmuch_config_close (config);
	return notmuch_setup_command (ctx, 0, NULL);
    }

    /* Notmuch is already configured, but is there a database? */
    db_path = talloc_asprintf (ctx, "%s/%s",
			       notmuch_config_get_database_path (config),
			       ".notmuch");
    if (stat (db_path, &st)) {
	notmuch_config_close (config);
	if (errno != ENOENT) {
	    fprintf (stderr, "Error looking for notmuch database at %s: %s\n",
		     db_path, strerror (errno));
	    return 1;
	}
	printf ("Notmuch is configured, but there's not yet a database at\n\n\t%s\n\n",
		db_path);
	printf ("You probably want to run \"notmuch new\" now to create that database.\n\n"
		"Note that the first run of \"notmuch new\" can take a very long time\n"
		"and that the resulting database will use roughly the same amount of\n"
		"storage space as the email being indexed.\n\n");
	return 0;
    }

    printf ("Notmuch is configured and appears to have a database. Excellent!\n\n"
	    "At this point you can start exploring the functionality of notmuch by\n"
	    "using commands such as:\n\n"
	    "\tnotmuch search tag:inbox\n\n"
	    "\tnotmuch search to:\"%s\"\n\n"
	    "\tnotmuch search from:\"%s\"\n\n"
	    "\tnotmuch search subject:\"my favorite things\"\n\n"
	    "See \"notmuch help search\" for more details.\n\n"
	    "You can also use \"notmuch show\" with any of the thread IDs resulting\n"
	    "from a search. Finally, you may want to explore using a more sophisticated\n"
	    "interface to notmuch such as the emacs interface implemented in notmuch.el\n"
	    "or any other interface described at http://notmuchmail.org\n\n"
	    "And don't forget to run \"notmuch new\" whenever new mail arrives.\n\n"
	    "Have fun, and may your inbox never have much mail.\n\n",
	    notmuch_config_get_user_name (config),
	    notmuch_config_get_user_primary_email (config));

    notmuch_config_close (config);

    return 0;
}

int
main (int argc, char *argv[])
{
    void *local;
    command_t *command;
    unsigned int i;

    local = talloc_new (NULL);

    g_mime_init (0);

    if (argc == 1)
	return notmuch (local);

    if (STRNCMP_LITERAL (argv[1], "--help") == 0)
	return notmuch_help_command (NULL, 0, NULL);

    if (STRNCMP_LITERAL (argv[1], "--version") == 0) {
	printf ("notmuch " STRINGIFY(NOTMUCH_VERSION) "\n");
	return 0;
    }

    for (i = 0; i < ARRAY_SIZE (commands); i++) {
	command = &commands[i];

	if (strcmp (argv[1], command->name) == 0)
	    return (command->function) (local, argc - 2, &argv[2]);
    }

    fprintf (stderr, "Error: Unknown command '%s' (see \"notmuch help\")\n",
	     argv[1]);

    talloc_free (local);

    return 1;
}
